### PR TITLE
WIP: Copy environ variable to FastCGI environ

### PR DESCRIPTION
**Pull request by geographika**
More of a proof-of-concept than a pull request. 

This gets the `CURL_CA_BUNDLE` value from the environment prior to `FCGI_Accept`, and then puts it into the FastCGI environment so it can be accessed later:

https://github.com/MapServer/MapServer/blob/cde09201ad2d469325bc911256513c5b9de78694/maphttp.c#L474

This works ok, but there are lots of MapServer environment variables. Some list/loop could be used?
Alternatively maybe adding the variables to the `mapserv->map` object would make more sense (and perhaps the same approach could be reused as part of https://github.com/sdlime/mapserver/wiki/MapServer-8.0-Config-File).

### CODE REVIEW

**Comment by rouault** [comment]
> Some list/loop could be used?

yep, declare "extern char **environ;" close to headers, iterate over it to make a copy of it, and then putenv in the fastcgi loop. But I would have expected the fastcgi code to do that for you...

File: mapserv.c [hunk 1]
```diff
@@ -239,12 +238,33 @@ int main(int argc, char *argv[])
...
 
+  // make a copy of the original environment
+  int i = 0;
+  int j;
+  char **env = environ;
+
+  while (env[i] != NULL)
+      i++;
+
+  char **envCopy = malloc(sizeof(char*) * i + 1);
```
Comment (rouault) [inline mapserv.c:249]: no need for +1 here:
```suggestion
  char **envCopy = malloc(sizeof(char*) * num_env_var);
```

File: mapserv.c [hunk 2]
```diff
@@ -239,12 +238,33 @@ int main(int argc, char *argv[])
...
+
+  for (j = 0; j < i; j++) {
+      envCopy[j] = malloc(strlen(env[j]));
```
Comment (rouault) [inline mapserv.c:252]: but here you need it
```suggestion
      envCopy[i] = malloc(strlen(env[i])+1);
```

File: mapserv.c [hunk 3]
```diff
@@ -239,12 +238,33 @@ int main(int argc, char *argv[])
   signal( SIGTERM, msCleanupOnSignal );
 #endif
 
+  // make a copy of the original environment
+  int i = 0;
+  int j;
```
Comment (rouault) [inline mapserv.c:243]:
```suggestion
```

File: mapserv.c [hunk 4]
```diff
@@ -239,12 +238,33 @@ int main(int argc, char *argv[])
...
+  char **env = environ;
+
+  while (env[i] != NULL)
+      i++;
+
+  char **envCopy = malloc(sizeof(char*) * i + 1);
+
+  for (j = 0; j < i; j++) {
```
Comment (rouault) [inline mapserv.c:251]:
```suggestion
  for (int i = 0; i < num_env_var; i++) {
```

File: mapserv.c [hunk 5]
```diff
@@ -239,12 +238,33 @@ int main(int argc, char *argv[])
   signal( SIGTERM, msCleanupOnSignal );
 #endif
 
+  // make a copy of the original environment
+  int i = 0;
```
Comment (rouault) [inline mapserv.c:242]:
```suggestion
  int num_env_var = 0;
```

File: mapserv.c [hunk 6]
```diff
@@ -239,12 +238,33 @@ int main(int argc, char *argv[])
...
+  int j;
+  char **env = environ;
+
+  while (env[i] != NULL)
```
Comment (rouault) [inline mapserv.c:246]:
```suggestion
  while (env[num_env_var] != NULL)
```

File: mapserv.c [hunk 7]
```diff
@@ -239,12 +238,33 @@ int main(int argc, char *argv[])
...
+      i++;
```
Comment (rouault) [inline mapserv.c:247]:
```suggestion
      num_env_var++;
```

File: mapserv.c [hunk 8]
```diff
@@ -239,12 +238,33 @@ int main(int argc, char *argv[])
...
+
+  char **envCopy = malloc(sizeof(char*) * i + 1);
+
+  for (j = 0; j < i; j++) {
+      envCopy[j] = malloc(strlen(env[j]));
+      strcpy(envCopy[j], env[j]);
```
Comment (rouault) [inline mapserv.c:253]:
```suggestion
      strcpy(envCopy[i], env[i]);
```

File: mapserv.c [hunk 9]
```diff
@@ -239,12 +238,33 @@ int main(int argc, char *argv[])
...
+
 #ifdef USE_FASTCGI
   msIO_installFastCGIRedirect();
 
   /* In FastCGI case we loop accepting multiple requests.  In normal CGI */
   /* use we only accept and process one request.  */
   while( FCGI_Accept() >= 0 ) {
+
+      // copy across all environ variables to the request environment
+      for (j = 0; j < i; j++) {
```
Comment (rouault) [inline mapserv.c:264]:
```suggestion
      for (int i = 0; i < num_env_var; i++) {
```

File: mapserv.c [hunk 10]
```diff
@@ -239,12 +238,33 @@ int main(int argc, char *argv[])
...
+          putenv(envCopy[j]);
```
Comment (rouault) [inline mapserv.c:265]:
```suggestion
          putenv(envCopy[i]);
```

**Review by rouault — COMMENTED** [review]
According to https://www.mkssoftware.com/docs/man5/environ.5.asp,
you should do
```
#include <unistd.h>
extern char **environ; 
```
Windows has likely not unistd.h however. According to https://docs.microsoft.com/en-us/cpp/c-runtime-library/environ-wenviron?view=msvc-140, you may need to use _environ

**Comment by geographika** [comment]
Thanks @rouault for the code review! I've added in the suggestions. I'm fairly sure some variables/memory needs to be freed at some stage but not sure where..

I'm still not sure if it is a good idea to go with this approach. Looking at the FastCGI docs, as I understand it, `getenv` *shouldn't* work on other web servers once in that loop. The FastCGI code uses `fcgi - 2.4.2` from [GISInternals](https://www.gisinternals.com/packageinfo.php?file=release-1928-x64-gdal-mapserver.zip), which is the latest version. Do Linux builds use the same?

Overwriting the env with all variables seems dangerous, but to check every key to see if it exists seems laborious. Messing with `PATH` etc. could cause further issues. 

Creating a config object outside the FastCGI loop may be better?

**Comment by rouault** [comment]
It looks like the approach of modifying the environment in combination with FastCGI is a dead-end
The doc of FCGI_Accept() mentions:
```
 * Side effects:
 *      If the application was invoked as a FastCGI server,
 *      and this is not the first call to this procedure,
 *      FCGI_Accept first performs the equivalent of FCGI_Finish.
 *
 *      On every call, FCGI_Accept accepts the new request and
 *      reads the FCGI_PARAMS stream into an environment array,
 *      i.e. a NULL-terminated array of strings of the form
 *      ``name=value''.  It assigns a pointer to this array
 *      to the global variable environ, used by the standard
 *      library function getenv.  It creates new FCGI_FILE *s
 *      representing input from the HTTP server, output to the HTTP
 *      server, and error output to the HTTP server, and assigns these
 *      new files to stdin, stdout, and stderr respectively.
 *
 *      DO NOT mutate or retain pointers to environ or any values
 *      contained in it (e.g. to the result of calling getenv(3)),
 *      since these are freed by the next call to FCGI_Finish or
 *      FCGI_Accept.  In particular do not use setenv(3) or putenv(3)
 *      in conjunction with FCGI_Accept.
```

**Comment by geographika** [comment]
@rouault - you added some good information in https://github.com/MapServer/MapServer/issues/6167#issuecomment-706116174 as part of #6167 - maybe `getenv` does cause issues in other environs/web servers?

**Comment by rouault** [comment]
ok, so for Apache / FCGI, after experimentations, I have now a good idea of what happens:
- if you add in your apache conf ``SetEnv FOO my_value`` (or ``PassEnv FOO`` and FOO is for example set in /etc/apache2/envvars), then getenv("FOO") only works *after* FCGI_Accept, and is added after the per-request CGI environment variables
- if you add ``FcgidInitialEnv FOO my_value``, then getenv("FOO") only works *before* FCGI_Accept. This is what you'd typically need for MS_DEBUGLEVEL which is read before the fcgi loop. In addition to what you explicitly pass that way, apache also seems to propagate automatically PATH and LD_LIBRARY_PATH

**Comment by sdlime** [comment]
> ok, so for Apache / FCGI, after experimentations, I have now a good idea of what happens:
> 
> * if you add in your apache conf `SetEnv FOO my_value` (or `PassEnv FOO` and FOO is for example set in /etc/apache2/envvars), then getenv("FOO") only works _after_ FCGI_Accept, and is added after the per-request CGI environment variables
> * if you add `FcgidInitialEnv FOO my_value`, then getenv("FOO") only works _before_ FCGI_Accept. This is what you'd typically need for MS_DEBUGLEVEL which is read before the fcgi loop. In addition to what you explicitly pass that way, apache also seems to propagate automatically PATH and LD_LIBRARY_PATH

So I interpret this as Apache CGI and FastCGI can access the environment vars as intended, correct?

**Comment by rouault** [comment]
> So I interpret this as Apache CGI and FastCGI can access the environment vars as intended, correct?

yes, but you have to be careful and rather knowledgeable of mapserver internals to know if you need to define a per-request var or an initial setting one.

**Comment by geographika** [comment]
> 
> 
> ok, so for Apache / FCGI, after experimentations, I have now a good idea of what happens:
> 
>     * if you add in your apache conf `SetEnv FOO my_value` (or `PassEnv FOO` and FOO is for example set in /etc/apache2/envvars), then getenv("FOO") only works _after_ FCGI_Accept, and is added after the per-request CGI environment variables
> 
>     * if you add `FcgidInitialEnv FOO my_value`, then getenv("FOO") only works _before_ FCGI_Accept. This is what you'd typically need for MS_DEBUGLEVEL which is read before the fcgi loop. In addition to what you explicitly pass that way, apache also seems to propagate automatically PATH and LD_LIBRARY_PATH

Thanks for that clarification. It looks like IIS FastCGI only supports the equivalent of `FcgidInitialEnv FOO my_value` with no possibility I can find for setting anything after FCGI_Accept. 

I think this approach is a dead-end in terms of trying to fix IIS and a config approach will be the cleaner/more robust option for setting global MapServer parameters, so closing. Thanks for the code suggestions though @rouault - my C has improved, as has my appreciation for how much easier things are in Python..

**Comment by rouault** [comment]
> with no possibility I can find for setting anything after FCGI_Accept.

so MS_MAP_PATTERN and the like can't be used with IIS in FastCGI, right ?

> my C has improved

that's what I'm slowly converting what I can to C++. std::vector<std::string> is less unfriendly than raw char** :-)

**Comment by geographika** [comment]
> so MS_MAP_PATTERN and the like can't be used with IIS in FastCGI, right ?

I've not been able to, and I've gone through pretty much any IIS/FastCGI doc I've been able to find.
It is however pretty trivial to filter querystring parameters to have the same result. 
The one environ setting that doesn't have a workaround is `CURL_CA_BUNDLE`. 

Is it worth adding this as a `CONFIG` option to a Mapfile and checking with `msGetConfigOption` as well as the environ, or waiting til 8.0?

**Comment by sdlime** [comment]
Could we store MS_MAP_PATTERN, etc... using CPLSetConfigOption() before FCGI_accept() and use that as a simple store in the meantime? Then use CPLGetConfigOption() in place of (or in addition to) getenv(). So no putenv() in the mix.

**Comment by rouault** [comment]
> Could we store MS_MAP_PATTERN, etc... using CPLSetConfigOption() before FCGI_accept() and use that as a simple store in the meantime? Then use CPLGetConfigOption() in place of (or in addition to) getenv(). So no putenv() in the mix.

That should work. At least for anything under MapServer and GDAL codebases (but not for PROJ's own env variables typically)

**Comment by sdlime** [comment]
What GDAL versions so CPLSetConfigOption()/CPLGetConfigOption() apply to, has it been around awhile?

**Comment by rouault** [comment]
> What GDAL versions so CPLSetConfigOption()/CPLGetConfigOption() apply to, has it been around awhile?

Not sure I was born yet when it was there :-) Anyway Mapserver already uses them : https://github.com/MapServer/MapServer/search?q=CPLSetConfigOption

**Comment by sdlime** [comment]
Logical to take advantage then.

**Comment by sdlime** [comment]
Perhaps search for getenv() and use that as a guide for what to grab - the MS_* stuff and things like SERVER_NAME, SERVER_PORT, etc... Then switch to CPLGetConfigOption() as much as possible. I'm assuming the calls in cgiutil.c would remain as would some other areas. Might be ok for both CGI and FCGI.

**Pull request closed without merging**